pkg/aws/ec2_helpers.go

```diff
@@ -134,10 +137,48 @@ func (ec AwsEc2DynamicConfig) configureInstance(name string, instanceTag string,
 		}},
 		InstanceInitiatedShutdownBehavior: types.ShutdownBehaviorTerminate,
 		InstanceMarketOptions:             instanceMarketOpts,
-		TagSpecifications:                 []types.TagSpecification{{ResourceType: types.ResourceTypeInstance, Tags: instanceTags}},
+		TagSpecifications: []types.TagSpecification{
+			{ResourceType: types.ResourceTypeInstance, Tags: instanceTags},
+		},
 	}
 }
 
+// findInstancesWithoutTaskRuns iterates over instances retrieved from the ec cloud and returns a list of those that
+// are associated with a non-existing Tekton TaskRun. Each instance should have a tag with the associated TaskRun's
+// namespace and name. This is compared to existingTaskRuns, which is a map of namespaces to a list of TaskRuns in
+// that namespace, to determine if this instance's TaskRun still exists.
+func (ec AWSEc2DynamicConfig) findInstancesWithoutTaskRuns(log logr.Logger, reservations []types.Reservation, existingTaskRuns map[string][]string) []string {
+	var instancesWithoutTaskRuns []string
+
+	// Iterate over all VM instances
+	for _, reservation := range reservations {
+		for _, instance := range reservation.Instances {
+			// Iterate over this instance's tags
+			for _, instanceTag := range instance.Tags {
+				if *instanceTag.Key != cloud.TaskRunTagKey {
+					continue
+				}
+				err := cloud.ValidateTaskRunId(*instanceTag.Key)
+				if err != nil {
+					log.Error(err, "invalid TaskRun ID; appending to list anyway...", "instanceID", *instance.InstanceId)
+					instancesWithoutTaskRuns = append(instancesWithoutTaskRuns, *instance.InstanceId)
+					break
+				}
+
+				// Try to find this instance's TaskRun
+				taskRunInfo := strings.Split((*instanceTag.Value), ":")
+				taskRunNamespace, taskRunName := taskRunInfo[0], taskRunInfo[1]
+				taskRuns, ok := existingTaskRuns[taskRunNamespace]
+				// Add the VM instance to the unassociated list if the TaskRun namespace or TaskRun does not exist
+				if !ok || !slices.Contains(taskRuns, taskRunName) {
+					instancesWithoutTaskRuns = append(instancesWithoutTaskRuns, *instance.InstanceId)
+				}
```

Comment:
If we reach here, we can `break`; no need to look at the other tags, right? 

It may be worth using something like `slices.IndexFunc` to remove this for loop and simplify the code a bit more. We could remove some break/continue